src/blocking-elements.ts

```diff
@@ -0,0 +1,404 @@
+
+interface Inertable extends HTMLElement {
+  inert?: boolean;
+}
+
+interface InternalState {
+  [_siblingsToRestore]: Set<MaybeHasInternalState>;
+  [_parentMO]: MutationObserver;
+}
+interface HasInternalState extends Inertable, InternalState {}
+interface MaybeHasInternalState extends Inertable, Partial<InternalState> {}
+
+/**
+ * `BlockingElements` manages a stack of elements that inert the interaction
+ * outside them. The top element is the interactive part of the document.
+ * The stack can be updated with the methods `push, remove, pop`.
+ */
+class BlockingElements {
+  /**
+   * The blocking elements.
+   */
+  private[_blockingElements]: MaybeHasInternalState[] = [];
+
+  /**
+   * Used to keep track of the parents of the top element, from the element
+   * itself up to body. When top changes, the old top might have been removed
+   * from the document, so we need to memoize the inerted parents' siblings
+   * in order to restore their inerteness when top changes.
+   */
+  private[_topElParents]: HasInternalState[] = [];
+
+  /**
+   * Elements that are already inert before the first blocking element is
+   * pushed.
+   */
+  private[_alreadyInertElements] = new Set<MaybeHasInternalState>();
+
+  /**
+   * Call this whenever this object is about to become obsolete. This empties
+   * the blocking elements
+   */
+  destructor() {
+    // Restore original inertness.
+    this[_restoreInertedSiblings](this[_topElParents]);
+    // Note we don't want to make these properties nullable on the class, since
+    // then we'd need non-null casts in many places. Calling a method on a
+    // BlockingElements instance after calling destructor will result in an
+    // exception.
+    const nullable = this as unknown as {
+      [_blockingElements]: null;
+      [_topElParents]: null;
+      [_alreadyInertElements]: null;
+    };
+    nullable[_blockingElements] = null;
+    nullable[_topElParents] = null;
+    nullable[_alreadyInertElements] = null;
+  }
+
+  /**
+   * The top blocking element.
+   */
+  get top(): HTMLElement|null {
+    const elems = this[_blockingElements];
+    return elems[elems.length - 1] || null;
+  }
+
+  /**
+   * Adds the element to the blocking elements.
+   */
+  push(element: HTMLElement): void {
+    if (!element || element === this.top) {
+      return;
+    }
+    // Remove it from the stack, we'll bring it to the top.
+    this.remove(element);
+    this[_topChanged](element);
+    this[_blockingElements].push(element);
+  }
+
+  /**
+   * Removes the element from the blocking elements. Returns true if the element
+   * was removed.
+   */
+  remove(element: HTMLElement): boolean {
+    const i = this[_blockingElements].indexOf(element);
+    if (i === -1) {
+      return false;
+    }
+    this[_blockingElements].splice(i, 1);
+    // Top changed only if the removed element was the top element.
+    if (i === this[_blockingElements].length) {
+      const top = this.top;
+      if (top !== null) {
```

Comment:
this check is not needed, _topChanged allows for `top === null`:

```js
this[_topChanged](this.top);
```

Reply:
Done.